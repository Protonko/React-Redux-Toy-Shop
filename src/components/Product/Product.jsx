import React from 'react';
import PropTypes from 'prop-types';
import placeholder from '../../assets/images/placeholder.jpg';
import {CountButton} from '../Common/Buttons/CountButton';

export const Product = ({image = placeholder, title, price, sale, count}) => {
  const priceProduct = sale ? Math.round(price * sale) : price;

  return (
    <div className="product">
      <div className="product__info">
        <div className="product__image" style={{backgroundImage: `url(${image})`} }/>

        <div className="product__data">
          <div className="product__data-name">{title}</div>
          <div className="product__data-price">${priceProduct}</div>
        </div>
      </div>

      <div className="product__counter">
        <CountButton isMinus={true} />
        <div className="product__counter-number">
          {count}
        </div>
        <CountButton isMinus={false} />
      </div>
    </div>
  );
};

Product.propTypes = {
  image: PropTypes.string,
  title: PropTypes.string,
  price: PropTypes.number,
  sale: PropTypes.number,
  count: PropTypes.number,
};
