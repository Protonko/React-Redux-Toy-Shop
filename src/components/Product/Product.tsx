import React, {FC} from 'react';
import Dotdotdot from 'react-dotdotdot';
import placeholder from 'assets/images/placeholder.jpg';
import {IProduct} from 'models/interfaces';
import {Button} from 'components/Common/Buttons/Button';

interface IProductProps {
  product: IProduct,
  removeFromCart: (product: IProduct) => any, // TODO - убрать
}

export const Product: FC<IProductProps> = ({
  product,
  removeFromCart,
}) => {
  const {image = placeholder, title, price, sale} = product;
  const priceProduct: number = sale ? Math.round(price * sale) : price;

  return (
    <div className="product">
      <div className="product__info">
        <div className="product__image" style={{backgroundImage: `url(${image})`} }/>

        <div className="product__data">
          <Dotdotdot clamp={1}>
            <div className="product__data-name">{title}</div>
          </Dotdotdot>
          <div className="product__data-price">${priceProduct}</div>
        </div>
      </div>

      <div className="product__actions">
        <ul className="product__actions-list list list--reset">
          <li className="product__actions-item">
            <Button
              title="Remove"
              onClick={() => removeFromCart(product)}
              isRemoved={true}
              classNamesAdditional={['product__action']}
            />
          </li>
        </ul>
      </div>
    </div>
  );
};
