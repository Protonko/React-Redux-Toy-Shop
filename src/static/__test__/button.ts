export const ELEMENT_SELECTORS_BUTTON = {
  button: 'button',
  testButton: 'test-button',
  minusButton: 'button--count-minus',
}
