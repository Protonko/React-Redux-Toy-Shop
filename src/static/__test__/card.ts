// types
import {IProduct} from 'models/interfaces';

export const product: IProduct = {
  id: 0,
  category: 'category',
  title: 'title',
  price: 10,
  sale: 0.9,
  rating: 5,
  checked: true,
};

export const product1: IProduct = {
  id: 1,
  category: 'category 1',
  title: 'title 1',
  price: 1,
  sale: 0.5,
  rating: 3,
  checked: true,
}

export const products: Array<IProduct> = [product, product1];
export const selectedProducts: Array<IProduct> = [product];
