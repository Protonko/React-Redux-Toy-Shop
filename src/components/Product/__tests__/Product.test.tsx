// types
import {IProps} from 'components/Product/Product';

import React from 'react';
import {shallow} from 'enzyme';
import {product} from 'static/__test__/card';
import {Product} from 'components/Product/Product';

const PRICE_DEFAULT = 10;
const SALE = 0.5;
const PRICE_WITH_SALE = 5;
const PRICE_PRODUCT_VALUE = '$9';
const ELEMENT_SELECTORS = {
  price: 'product__data-price',
}

const shallowCardComponent = (props: IProps) => (
  shallow(<Product {...props} />)
);

describe('Product', () => {
  let props: IProps;
  let price: number;
  let sale: number | void;

  beforeEach(() => {
    props = {
      product,
      removeFromCart: () => {},
    };
    price = PRICE_DEFAULT;
  });

  it('Match snapshot', () => {
    const component = shallowCardComponent(props);

    expect(component).toMatchSnapshot();
  });

  it('Should price be PRICE_DEFAULT === 10', () => {
    const priceProduct = sale ? Math.round(price * sale) : price;

    expect(priceProduct).toBe(PRICE_DEFAULT);
  });

  it('Should return price with sale', () => {
    sale = SALE;
    const priceProduct = sale ? Math.round(price * sale) : price;

    expect(priceProduct).toBe(PRICE_WITH_SALE);
  });

  it('Should return price with sale', () => {
    const productComponent = shallowCardComponent(props);
    const productPrice = productComponent.find(`.${ELEMENT_SELECTORS.price}`);
    const priceValue = productPrice.text();

    expect(priceValue).toBe(PRICE_PRODUCT_VALUE);
  });
});
