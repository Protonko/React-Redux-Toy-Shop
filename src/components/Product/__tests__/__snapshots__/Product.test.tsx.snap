// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Match snapshot 1`] = `
<div
  className="product"
>
  <div
    className="product__info"
  >
    <div
      className="product__image"
      style={
        Object {
          "backgroundImage": "url([object Object])",
        }
      }
    />
    <div
      className="product__data"
    >
      <Dotdotdot
        clamp={1}
        tagName="div"
        truncationChar="â€¦"
        useNativeClamp={false}
      >
        <div
          className="product__data-name"
        >
          title
        </div>
      </Dotdotdot>
      <div
        className="product__data-price"
      >
        $
        9
      </div>
    </div>
  </div>
  <div
    className="product__actions"
  >
    <ul
      className="product__actions-list list list--reset"
    >
      <li
        className="product__actions-item"
      >
        <Component
          classNamesAdditional={
            Array [
              "product__action",
            ]
          }
          isRemoved={true}
          onClick={[Function]}
          title="Remove"
        />
      </li>
    </ul>
  </div>
</div>
`;
